### DISCUSSION TITLE
Collection expressions vs readonly

### DISCUSSION THREAD

**Discussion by fandrei** [dotnet/csharplang#8889] [Language Ideas] [answered]
The following code works:
1.
```C#

	class Test
	{
		public readonly List<int> List = [];
	}

	static readonly Test Obj = new()
	{
		List = { 0 },
	};
```
and this code works too:
2.
```C#
	static readonly List<int> Vals = [0];
```
However, this code doesn't compile:
3.
```C#
	static readonly Test Obj1 = new()
	{
		List = [0],
	};
```
Shouldn't it work in this case, too?

**Comment by CyrusNajmabadi** [comment, marked as answer]
> Shouldn't it work in this case, too?

No.  The list is readonly, so it cannot be assigned.  

To your examples:

```c#
	static readonly Test Obj = new()
	{
		List = { 0 },
	};
```

This is not an assignment of hte list.  This is translated to:

```c#
Test __test = new Test();
__test.List.Add(0);
Obj = __test;
```

This is fine as the list is not assigned to, it just has a value added to it.

```c#
	static readonly List<int> Vals = [0];
```

This is the initializer of a readonly field, which is always legal.  A readonly field can always be initialized with an iniitalizer, or by an assignment in the constructor of the type owning the readonly field.

**Reply by fandrei** [reply]
- The collection in 2. also is readonly, but this code works.
- It can call AddRange, doesn't it?

**Reply by CyrusNajmabadi** [reply]
> The collection in 2. also is readonly, but this code works.

You are in the field initializer in one case.  In the other you are not in the field initializer.  A value for a readonly field can only be provided in the initializer, or in the constructor.  After that point, the field is now readonly and cannot be assigned (that's what it means to be readonly).

**Reply by CyrusNajmabadi** [reply]
> It can call AddRange, doesn't it?

No.  It creates a collection and attempts to assign it.  As the field is readonly, it cannot be assigned in a collection initializer.

**Reply by fandrei** [reply]
> It creates a collection and attempts to assign it. As the field is readonly, it cannot be assigned in a collection initializer.

So why doesn't it be a bit smarter and call AddRange in this case, too? The same way as it did when using old syntax?

**Reply by jnm2** [reply]
~~@fandrei Someone would have to make the case that this new codegen is an improvement. Sometimes what sounds simpler at first doesn't actually play out the way you'd expect. For example, AddRange requires an allocation of a collection around the items that are being added, and Add does not.~~

**Reply by fandrei** [reply]
@jnm2 I don't see why this would be too difficult to do. It just doesn't make any sense to me that this shiny brand new initialization syntax is inferior to old syntax, in some aspects.

**Reply by 333fred** [reply]
> Someone would have to make the case that this new codegen is an improvement. 

It's not about codegen. It's about semantics. One thing means "create a new collection". The other means "add to an existing collection". The first thing that would need to happen is convincing the LDM that it's a good idea for the semantic meaning of the code to change based on whether the field/property can be assigned. That seems unlikely to happen, to put it mildly.

**Reply by jnm2** [reply]
Ah, you're right, I thought it was a codegen request for `= { a, b, c }` to use AddRange but I misread.

Agreed. `= { a, b, c }` and `= new List<int> { a, b, c }` do not mean the same thing. And `= [a, b, c]` already means the latter, not the former. So, it can't mean the former.

**Reply by CyrusNajmabadi** [reply]
It would also be a breaking change. For people that want the assignment, calling .Add will break (and quite possibly throw). We would truly break programs changing the making meaning here

**Reply by fandrei** [reply]
In this cases, how does `	public Test() { List = [0]; }` work?

**Reply by CyrusNajmabadi** [reply]
Because you're in the constructor, as I mentioned.  That's precisely where a readonly field can be assigned. This is also unrelated to collection expressions :-)

**Reply by CyrusNajmabadi** [reply]
> @jnm2 I don't see why this would be too difficult to do. It just doesn't make any sense to me that this shiny brand new initialization syntax is inferior to old syntax, in some aspects.

{} It's the syntax we use for initialization (mutating a variable that already exists).  `= expr` is for assigning a new value.  They are fundamentally different semantics based on different needs. 

The former is when the value already exists and the calling code wants to change it. The latter is for assigning a new value. These are fundamentally different concerns, so they have differing syntax.

**Reply by fandrei** [reply]
@CyrusNajmabadi
> {} It's the syntax we use for initialization (mutating a variable that already exists)

Doesn't `= {}` mean allocation of the new variable in some cases?

**Reply by jnm2** [reply]
No, it never allocates a new collection. You would write `= new() {` or something similar if you wanted that, and it would only be permitted in locations where you can assign to the property or field.

**Reply by jnm2** [reply]
Oh, `= {` does allocate a new array in a variable initializer specifically, though not in an object initializer or in an ordinary assignment statement. This is only for arrays, and only in a variable initializer.

**Reply by fandrei** [reply]
@jnm2 what about `int[] vals = { 0 }`? PS I see that you wrote about that already. So, sometimes it does that.

**Reply by CyrusNajmabadi** [reply]
In an object initializer, `{}` never allocates.  This includes for arrays.  It's always .Add.  You'll note you cannot use `= { ... }` to initialize an array in an object initializer.

**Reply by jnm2** [reply]
You can think of that as special syntax, maybe even as legacy syntax, but it doesn't really come into the question about object initializers. We still have `{ CollectionMember = { a, b, c } }` and `{ CollectionMember = new List<int> { a, b, c } }`, which differ in meaning in a very important way: are you appending a, b, and c to an existing collection, or creating and assigning a new collection?

And `{ CollectionMember = [a, b, c] }` already means one, and not the other. It can't mean both at once, and like Fred said, having it switch between meanings (append versus replace) based on distant code is highly unpalatable.

**Reply by fandrei** [reply]
> In an object initializer, `{}` never allocates

I see. But in field and local variable initializers, it does. Special cases == more confusion.

**Reply by CyrusNajmabadi** [reply]
That decision was made in c# 3.0, almost 20 years ago.  It's just something we'll have to live with because we cannot change it (for all teh reasons it would break that we've given above).

**Comment by CyrusNajmabadi** [comment]
Note: this is unrelated to collection expression.  The following are illegal as well:

```c#
	static readonly Test Obj1 = new()
	{
		List = new(),
	};

	static readonly Test Obj1 = new()
	{
		List = new List<int>(),
	};

	static readonly Test Obj1 = new()
	{
		List = null,
	};

	static readonly Test Obj1 = new()
	{
		List = someOtherList,
	};
```

All of these are illegal, not because a collection expr was used, but because the field is readonly.  You cannot assign it outside of hte field-initializer/constructor.

**Reply by fandrei** [reply]
It doesn't have to be an assignment. Add/AddRange would do just fine, like in other cases.

**Reply by CyrusNajmabadi** [reply]
`= expr` *is* an assignment, where the expectation is that the RHS overwrites.  `= { ... }` is initialization, where the expectation is that the `{ ... }` adds.  Having `[...]` add would break expectations.  Consider something simple like:

```
new () { ListWithADefaultElementAlreadyInIt = [a, b, c] }
```

Previously this would give you just the elements `a, b, c`.  Now it would include the *existing* elements already in that collection and hte new elements.  This would break code.

**Reply by fandrei** [reply]
> Now it would include the _existing_ elements already in that collection and hte new elements. This would break code.

I get it that you guys are very reluctant to change any decisions, but as a theoretical exercise, what's the probability that existing code in real apps would break for this reason?

**Reply by CyrusNajmabadi** [reply]
> > Now it would include the _existing_ elements already in that collection and hte new elements. This would break code.
> 
> I get it that you guys are very reluctant to change any decisions, but as a theoretical exercise, what's the probability that existing code in real apps would break for this reason?

It would be high.  Assigning versus adding is very different semantics. For 20 years of code, something will certainly have taken a dependency on this

**Reply by jnm2** [reply]
Not only that, but what's the motivation to make a change? What is it you're trying to do that `= { 0 }` isn't sufficient for?

**Reply by fandrei** [reply]
> It would be high. Assigning versus adding is very different semantics. For 20 years of code, something will certainly have taken a dependency on this

`= []` is not 20 years old, it's very new. I'd estimate this probability as less than 100 cases worldwide.

**Reply by fandrei** [reply]
> Not only that, but what's the motivation to make a change? What is it you're trying to do that `= { 0 }` isn't sufficient for?

Just the principle of least surprise.

**Reply by jnm2** [reply]
The surprise isn't the problematic kind; you find out immediately that it isn't going to work, and you already know what does work instead. In my opinion, it would be a stretch to call this a "high astonishment factor." Sure, there's always going to be a series of minor surprises in any platform.

**Reply by jnm2** [reply]
> `= []` is not 20 years old, it's very new. I'd estimate this probability as less than 100 cases worldwide.

Leaving aside the estimate of one person on the internet (thanks for that 😄), making `= [` able to invisibly switch behavior would increase code review burden for all future decades, and would slow everyone down who sees it in an object initializer. You would then have to second-guess what it means every time you see it. The meaning would no longer be evident in the syntax. Whereas with `= {` versus `= new List<int> {`, you can see the `new` keyword right there in the syntax and know which one is happening. And by intentional design, `= [` always means "construct a new collection," so people have learned that it is equivalent in meaning to using `new`.

**Reply by fandrei** [reply]
@jnm2 it's just another technicality that users need to remember. Use ` = []` in most cases, but you have to use ` = {}` there. Except those cases, when you have to use ` = new() {}` or  ` = new() []` or ` = new MyList() {}`, etc.
For the result that you want to achieve, it simply doesn't matter what it does under the hood - creates a new collection, or adds to an empty one.

**Reply by jnm2** [reply]
It might not be adding to an empty one, which is why it does matter. The defaults for every collection on every new object may not be empty.

**Reply by fandrei** [reply]
> It might not be adding to an empty one, which is why it does matter. The defaults for every collection on every new object may not be empty.

Looks like we made a full circle. `= {}` does exactly that, and I don't remember anyone actually having problems because of that. :shrug:
And if you think that it's a problem for code reviews, I hope that you enjoy target-typed news :) 👿

**Reply by jnm2** [reply]
You've missed my point (originally Fred's), or you wouldn't think that I was saying there was any problem with `= {`.

**Reply by fandrei** [reply]
> You've missed my point (originally Fred's), or you wouldn't think that I was saying there was any problem with `= {`.

or you missed mine. :shrug:

**Reply by CyrusNajmabadi** [reply]
> Just the principle of least surprise.

It would be very surprising to me that a collection expression doesn't behave like a collection expression in this single case.

it would also be very surprising to me that an object initializer assignment doesn't assign, like it has since inception.  

There are two *very* surprising things you are suggesting.

**Reply by CyrusNajmabadi** [reply]
> = [] is not 20 years old, it's very new. I'd estimate this probability as less than 100 cases worldwide.

I don't see any value in breaking *anyone* here.  The behavior it has is what i would expect.  The behavior you are suggesting is quite surprising and out of line with both object initializers and how they've operated for 20 years, and how we want collection expressions to operate (and how they operate everywhere else).  So this would both break people, and be an undesirable change in multiple dimensions.  That's multiple strikes against it.

**Reply by CyrusNajmabadi** [reply]
> Looks like we made a full circle. = {} does exactly that, and I don't remember anyone actually having problems because of that.

No one is asking to change `{ }`, which has been spec'ed to work this way and which has worked this way *intentionally* for '20' years.  Again, it's intentional that we have two forms to describe two disparate operations within an object initializer.  There is `= { ... }` to *not* overwrite, but instead act *upon* an existing value.  and there is `= expr` to overwrite. 

`[...]` is an expression.  It produces a value, it would be surpising for this not to act as a value, and overwrite, like every other collecting producing expression out there.  `= { .. }` *is* the exception.  It exists entirely to act on an existing value.  No other expressions are intended for that purpose in an object initializer, and we would not want `[ ... ]` to do this as it would make it act entirely inappropriately.

**Reply by CyrusNajmabadi** [reply]
> = [] is not 20 years old, it's very new. I'd estimate this probability as less than 100 cases worldwide.

It would instantly break everyone who had done:

```c#
public record Anything(ImmutableArray<string> Values);

new Anything { Values = ["a", "b", "c"] }
```

This would now try adding to a non-initialized ImmutableArray, immediately throwing.  This would also be the case for *every* other collection type.  `List<string>` would throw as well, as calling .Add on null would break.

Every single consumer of collection expressions that intend to assign/initialize a value would now break.  

We are not makign this change.  Especially as it makes object-initializers less consistent with behavior around initialization or assignment of a property.

**Reply by CyrusNajmabadi** [reply]
Cases that immediately break if you move to `= [...]` calling 'Add' instead of assigning:

```c#
class C
{
    ImmutableArray<string> Values { get; set }
}

new C()
{
    Values = ["a"]; // this not throws, calling .Add on a 'default' immutable array.
}
```

or:

```c#
class C
{
    ImmutableArray<string> Values { get; set }  = ImmutableArray<string>.Empty;
}

new C()
{
    Values = ["a"]; // this now adds *nothing*, since calling .Add on ImmutableArray doesn't mutate the value.
}
```

Both of the above cases are legal C# *today* that works exactly how anyone would expect.  You get a new C instance whose 'Values' property is an array with a single element with it.  With your change both break.  one breaks by throwing, the other breaks by losing values.

This pattern does exist, and it's certainly more than 100 cases worldwide.

**Reply by fandrei** [reply]
> This pattern does exist, and it's certainly more than 100 cases worldwide.

I'm sure that it's not 20 years of code as you wrote before. This code wasn't even possible until very recently.

**Reply by CyrusNajmabadi** [reply]
> This code wasn't even possible until very recently.

And yet it has distinct and specific meanings, that code *will* depend on.  As shown above, if you change this meaning, you will break even *trivial* use cases.

**Reply by CyrusNajmabadi** [reply]
> I'm sure that it's not 20 years of code as you wrote before.

This was in reference to the expectations around *how* object initializers work and the dependencies people have on it.  Changing anything there is likely to break code.  Even if we limit this *just* to collection expressions, it is *still* definitely breaking as even basic cases like the above *which absolutely do exist* will now break.  

Breaking that code is *not* acceptable.  The meaning it has is:

a. what we want
b. what the ecosystem wants
c. what real world programs absolutely depend on

We will not break this.  It would be harmful.
